### PR TITLE
Parallel energy improvements

### LINKED ISSUE #219: Exclude duplicate atom pairs through cell loop
**Describe the feature / issue**
Cell-based energy and force calculation could move the excludeIgtJ check out to the Cells themselves, rather than checking each individual atom.

### PR DESCRIPTION
**Pull request by StephenSmith25**
This PR improves the distribution of work in the energy parallelisation. 

It uses a precomputed matrix of neighbours, which is distributed over the threads. It means more of the work is parallel. Hopefully should have better scaling. 
My timings:

```
Benchmark before threading:

TotalInterAtomicMolecular<Medium> 2436ms

Before this PR but with threading

TotalInterAtomicMolecular<Medium> 391ms

Now,

TotalInterAtomicMolecular<Medium> 147ms
```

Closes #219.

### CODE REVIEW

**Review by trisyoungs — CHANGES_REQUESTED** [review]
This is exactly the kind of thing I had in mind.

Couple of small comments, and one probably dumb question.

File: src/classes/energykernel.cpp [hunk 1]
```diff
@@ -567,22 +567,29 @@ double EnergyKernel::energy(const CellArray &cellArray, bool interMolecular, Pro
...
+                                                             auto *cellI = cellArray.cell(i);
+                                                             auto *cellJ = cellNeighbourArray[i][j].neighbour_;
+                                                             if (!cellJ)
+                                                                 return 0.0;
+                                                             auto mimRequired = cellNeighbourArray[i][j].requiresMIM_;
+                                                             runningTotal += energy(cellI, cellJ, mimRequired, true,
+                                                                                    interMolecular, subStrategy, performSum);
+                                                         }
+                                                     }
+                                                     return runningTotal;
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply src/classes/energykernel.cpp:590]: Yeah good point. But wasn't it doing this before? Maybe I'm missing something, but I looked at what it was doing before, which seemed to be:
- Loop over cells
- Call energy calculation of cell with itself
- Call energy calculation for its neighbours and mim neighbours. 

I'll rework it to account for the symmetry

[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/classes/energykernel.cpp:590]: Possibly, but that would have been wrong as well!  I'll have a check of this ahead of the TSC.

**Comment by StephenSmith25** [comment]
I've reworked how the energy is calculated for the cell array. It loops over the list of pairs (upper part of the matrix flattened as a 1d array) and finds the energy between the pairs of atoms in cells, rather than using the arrayIndex() method to omit repeated calculations. I've also pulled out the MPI distribution of the atoms in each pair of cells. I was thinking that the cells have already been distributed over MPI nodes and therefore we'd need to check if further distributing the atoms is beneficial.

**Comment by trisyoungs** [comment]
Apologies - got carried away with a fix and, apparently, I have push access to your branch!  Terribly bad etiquette.

Here's the result anyway - I traced the crash back to `EnergyKernel::energy(const CellArray &cellArray ...)` in which the cell neighbour pairs seemed to refer to bad (deleted) pointers.  I couldn't see an obvious reason for this, but did realise that we didn't need to use a pointer-based array for `Cell` at all.  The array/vector is created once and never changes, so a plain `std::vector<Cell>` created to the known required size does the trick.  This clears up the crash, and as a by-product enforces that all pointers to `Cell`s returned by `CellArray` must be `const`.  The only caveat to this last point is when an atom needs to check its position to see if it has moved to a different cell.  I've put in a non-`const` overload of one of the access functions to get around that.

See what you think!

And again, apologies for trampling over your PR...

**Comment by StephenSmith25** [comment]
Thanks for looking at that! I pushed one further commit to make everything build - The vec3 constructor is now explicit as, `cell(1)` was ambiguous as it could either be `cell(int)` or `cell(vec3<>) `as a single integer can be implicitly converted into a vector, which should be avoided. Should be ready for a re-review.

**Review by trisyoungs — APPROVED** [review]
:+1:

**Pull request merged**